File: apps/api/src/domain/ride-status.ts
import { InvalidTransitionError } from '../common/errors';

/**
 * The lifecycle of a single ride leg. Mirrors lib/domain/ride_status.dart —
 * the client keeps its copy so it can disable controls, but **this** one is
 * authoritative. A client cannot be trusted with a state machine.
 *
 * `delayed` is deliberately not a member. A driver stuck in traffic on the way
 * to pickup is still `driverEnRoute`; making delay a status would lose the
 * state it has to return to. Delay is a flag on the ride plus a RideEvent.
 */
export type RideStatus =
  | 'draft'
  | 'requested'
  | 'awaitingAssignment'
  | 'assigned'
  | 'driverAccepted'
  | 'driverEnRoute'
  | 'driverArrived'
  | 'passengerOnboard'
  | 'inProgress'
  | 'arrivedAtDestination'
  | 'completed'
  | 'canceled'
  | 'noShow'
  | 'reassignmentRequired';

export const RIDE_STATUSES: readonly RideStatus[] = [
  'draft',
  'requested',
  'awaitingAssignment',
  'assigned',
  'driverAccepted',
  'driverEnRoute',
  'driverArrived',
  'passengerOnboard',
  'inProgress',
  'arrivedAtDestination',
  'completed',
  'canceled',
  'noShow',
  'reassignmentRequired',
];

const TERMINAL = new Set<RideStatus>(['completed', 'canceled', 'noShow']);

/**
 * Whether the driver app may share location and the family app may show it.
 *
 * The single definition of "tracking is legal right now". Re-checked on every
 * inbound location point. Tracking begins when the driver sets off and ends the
 * moment the ride reaches a terminal state.
 */
const LOCATION_SHARING = new Set<RideStatus>([
  'driverEnRoute',
  'driverArrived',
  'passengerOnboard',
  'inProgress',
  'arrivedAtDestination',
]);

const PASSENGER_ONBOARD = new Set<RideStatus>(['passengerOnboard', 'inProgress']);

export function isTerminalRideStatus(status: RideStatus): boolean {
  return TERMINAL.has(status);
}

export function allowsLocationSharing(status: RideStatus): boolean {
  return LOCATION_SHARING.has(status);
}

export function passengerIsOnboard(status: RideStatus): boolean {
  return PASSENGER_ONBOARD.has(status);
}

/**
 * Allowed transitions. Anything not listed here is rejected.
 *
 * Cancellation is permitted from every state where the ride is still happening
 * — including after pickup, because the reasons for stopping one are rarely
 * convenient. The single exception is `arrivedAtDestination`: the passenger has
 * already been delivered, so there is nothing left to call off and the only
 * move is to complete.
 */
const ALLOWED: Record<RideStatus, readonly RideStatus[]> = {
  draft: ['requested', 'canceled'],
  requested: ['awaitingAssignment', 'canceled'],
  awaitingAssignment: ['assigned', 'reassignmentRequired', 'canceled'],
  assigned: ['driverAccepted', 'reassignmentRequired', 'canceled'],
  driverAccepted: ['driverEnRoute', 'reassignmentRequired', 'canceled'],
  driverEnRoute: ['driverArrived', 'reassignmentRequired', 'canceled'],
  driverArrived: ['passengerOnboard', 'noShow', 'canceled'],
  passengerOnboard: ['inProgress', 'canceled'],
  inProgress: ['arrivedAtDestination', 'canceled'],
  arrivedAtDestination: ['completed'],
  reassignmentRequired: ['assigned', 'canceled'],
  completed: [],
  canceled: [],
  noShow: [],
};

export function canTransitionRide(from: RideStatus, to: RideStatus): boolean {
  return ALLOWED[from]?.includes(to) ?? false;
}

export function allowedRideTransitions(from: RideStatus): readonly RideStatus[] {
  return ALLOWED[from] ?? [];
}

/** Throws unless the transition is legal. Every status change funnels here. */
export function assertRideTransition(from: RideStatus, to: RideStatus): void {
  if (!canTransitionRide(from, to)) {
    throw new InvalidTransitionError(from, to);
  }
}

File: lib/domain/ride_status.dart
import 'package:carebridge_client/carebridge_client.dart';

/// The lifecycle of a single ride leg.
///
/// A round trip is **two rides** sharing a `roundTripGroupId`, not one ride with
/// two legs: each leg is assigned, tracked, delayed, cancelled and priced
/// independently, and the return leg's pickup time is frequently unknown when
/// the outbound leg is booked.
///
/// Deviation from the original status list, deliberate: `delayed` is **not** a
/// status here. A delay does not replace where a ride is — a driver stuck in
/// traffic on the way to pickup is still `driverEnRoute` — and modelling it as a
/// status loses the state it must return to. Delay is carried on the ride as a
/// flag plus a `RideEvent`, so it can be raised and cleared from any active
/// state without corrupting the machine. See `Ride.isDelayed`.
enum RideStatus {
  draft,
  requested,
  awaitingAssignment,
  assigned,
  driverAccepted,
  driverEnRoute,
  driverArrived,
  passengerOnboard,
  inProgress,
  arrivedAtDestination,
  completed,
  canceled,
  noShow,
  reassignmentRequired;

  /// Plain-language label. Written for a worried adult child reading it on a
  /// phone, not for an operations console.
  String get label => switch (this) {
    RideStatus.draft => 'Draft',
    RideStatus.requested => 'Requested',
    RideStatus.awaitingAssignment => 'Finding a driver',
    RideStatus.assigned => 'Driver assigned',
    RideStatus.driverAccepted => 'Driver confirmed',
    RideStatus.driverEnRoute => 'Driver on the way',
    RideStatus.driverArrived => 'Driver has arrived',
    RideStatus.passengerOnboard => 'Picked up',
    RideStatus.inProgress => 'On the way to the clinic',
    RideStatus.arrivedAtDestination => 'Arrived at the clinic',
    RideStatus.completed => 'Completed',
    RideStatus.canceled => 'Canceled',
    RideStatus.noShow => 'No show',
    RideStatus.reassignmentRequired => 'Needs a new driver',
  };

  bool get isTerminal =>
      this == RideStatus.completed ||
      this == RideStatus.canceled ||
      this == RideStatus.noShow;

  /// Whether the driver app may share location, and the family app may show it.
  ///
  /// This is the single definition of "tracking is legal right now". The server
  /// re-checks it on every inbound location point; the client uses it to decide
  /// whether to start the location stream at all. Tracking begins when the
  /// driver sets off and ends the moment the ride reaches a terminal state.
  bool get allowsLocationSharing => const {
    RideStatus.driverEnRoute,
    RideStatus.driverArrived,
    RideStatus.passengerOnboard,
    RideStatus.inProgress,
    RideStatus.arrivedAtDestination,
  }.contains(this);

  /// Whether the patient is physically in the vehicle. Drives the family app's
  /// most-asked question and the "picked up safely" notification.
  bool get passengerIsOnboard =>
      const {RideStatus.passengerOnboard, RideStatus.inProgress}.contains(this);
}

/// Allowed transitions. Anything not listed here is rejected.
///
/// Cancellation is permitted from every state where the ride is still happening
/// — including after pickup, because the reasons for stopping one are rarely
/// convenient. The single exception is `arrivedAtDestination`: the passenger has
/// already been delivered, so there is nothing left to call off and the only
/// move is to complete.
const Map<RideStatus, Set<RideStatus>> _allowed = {
  RideStatus.draft: {RideStatus.requested, RideStatus.canceled},
  RideStatus.requested: {RideStatus.awaitingAssignment, RideStatus.canceled},
  RideStatus.awaitingAssignment: {
    RideStatus.assigned,
    RideStatus.reassignmentRequired,
    RideStatus.canceled,
  },
  RideStatus.assigned: {
    RideStatus.driverAccepted,
    RideStatus.reassignmentRequired,
    RideStatus.canceled,
  },
  RideStatus.driverAccepted: {
    RideStatus.driverEnRoute,
    RideStatus.reassignmentRequired,
    RideStatus.canceled,
  },
  RideStatus.driverEnRoute: {
    RideStatus.driverArrived,
    RideStatus.reassignmentRequired,
    RideStatus.canceled,
  },
  RideStatus.driverArrived: {
    RideStatus.passengerOnboard,
    RideStatus.noShow,
    RideStatus.canceled,
  },
  RideStatus.passengerOnboard: {RideStatus.inProgress, RideStatus.canceled},
  RideStatus.inProgress: {RideStatus.arrivedAtDestination, RideStatus.canceled},
  RideStatus.arrivedAtDestination: {RideStatus.completed},
  RideStatus.reassignmentRequired: {RideStatus.assigned, RideStatus.canceled},
  RideStatus.completed: {},
  RideStatus.canceled: {},
  RideStatus.noShow: {},
};

bool canTransitionRide(RideStatus from, RideStatus to) =>
    _allowed[from]?.contains(to) ?? false;

Set<RideStatus> allowedRideTransitions(RideStatus from) =>
    _allowed[from] ?? const {};

/// Throws [InvalidTransitionFailure] unless the transition is legal.
///
/// Every ride status change in the app funnels through here. Reaching it with an
/// illegal transition means a control was offered that should have been hidden
/// or disabled — a UI bug, surfaced loudly rather than silently persisted.
void assertRideTransition(RideStatus from, RideStatus to) {
  if (!canTransitionRide(from, to)) {
    throw InvalidTransitionFailure(from.name, to.name);
  }
}
